src/receive_imf.rs

```diff
@@ -2326,47 +2348,37 @@ async fn has_verified_encryption(
 
     let contact = Contact::get_by_id(context, from_id).await?;
 
-    for (to_addr, mut is_verified) in rows {
-        info!(
-            context,
-            "has_verified_encryption: {:?} self={:?}.",
-            to_addr,
-            context.is_self_addr(&to_addr).await
-        );
-        let peerstate = Peerstate::from_addr(context, &to_addr).await?;
-
-        // mark gossiped keys (if any) as verified
-        if mimeparser.gossiped_addr.contains(&to_addr.to_lowercase()) {
-            if let Some(mut peerstate) = peerstate {
-                // if we're here, we know the gossip key is verified:
-                // - use the gossip-key as verified-key if there is no verified-key
-                // - OR if the verified-key does not match public-key or gossip-key
-                //   (otherwise a verified key can _only_ be updated through QR scan which might be annoying,
-                //   see <https://github.com/nextleap-project/countermitm/issues/46> for a discussion about this point)
-                if !is_verified
-                    || peerstate.verified_key_fingerprint != peerstate.public_key_fingerprint
-                        && peerstate.verified_key_fingerprint != peerstate.gossip_key_fingerprint
-                {
-                    info!(context, "{} has verified {}.", contact.get_addr(), to_addr);
+    for (to_addr, is_verified) in rows {
+        if let Some(mut peerstate) = Peerstate::from_addr(context, &to_addr).await? {
+            // mark gossiped keys (if any) as verified
+            if mimeparser.gossiped_addr.contains(&to_addr.to_lowercase()) {
+                // If we're here, we know the gossip key is verified.
+                // - Use the gossip-key as verified-key if there is no verified-key
+                // - Store gossip key as secondary verified key if there is a verified key and
+                //   gossiped key is different.
+                //
+                // See <https://github.com/nextleap-project/countermitm/issues/46>
+                // and <https://github.com/deltachat/deltachat-core-rust/issues/4541> for discussion.
+                if !is_verified {
```

Comment:
Would be nice to switch `if` and `else` in order to get rid of the negation. If you do it, remember to also switch the two bullet points in the comment ("Use the gossip-key..." and "Store gossip key...").